src/django_pg_migration_tools/operations.py

```diff
@@ -226,5 +236,185 @@ def database_backwards(
+            return
+
+        index = self._get_index_for_constraint()
+        model = to_state.apps.get_model(app_label, self.model_name)
+        self.safer_create_index(
+            app_label=app_label,
+            schema_editor=schema_editor,
+            from_state=from_state,
+            to_state=to_state,
+            index=index,
+            model=model,
+            unique=True,
+        )
+
+        # Django doesn't have a handy flag "using=..." so we need to alter the
+        # SQL statement manually. We go from a SQL that looks like this:
+        #
+        # - ALTER TABLE "table" ADD CONSTRAINT "constraint" UNIQUE ("field")
+        #
+        # Into a SQL that looks like:
+        #
+        # - ALTER TABLE "table" ADD CONSTRAINT "constraint" UNIQUE USING INDEX "idx"
+        base_sql = str(self.constraint.create_sql(model, schema_editor))
+        alter_table_sql = base_sql.split(" UNIQUE")[0]
+        sql = f'{alter_table_sql} UNIQUE USING INDEX "{index.name}"'
+
+        # Run the sql to add the constraint inside a transaction. We have lock
+        # timeouts back in place after creating the index that would prevent
+        # this operation from running for too long if it's blocked by another
+        # query. Otherwise, this operation should actually be quite fast - if
+        # it's not blocked - since we have created the unique index in the
+        # previous step.
+        with transaction.atomic():
+            return schema_editor.execute(sql)
+
+    def database_backwards(
+        self,
+        app_label: str,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        from_state: migrations.state.ProjectState,
+        to_state: migrations.state.ProjectState,
+    ) -> None:
+        if not self._can_migrate(app_label, schema_editor, from_state):
+            return
+
+        if not self._constraint_exists(schema_editor):
+            # Nothing to delete.
+            return
+
+        # Drop the constraint in a transaction so that it can raise a lock
+        # timeout error if it takes too long. This transaction.atomic ctx
+        # manager is required because the "forward" part of the migration
+        # requires an atomic=False flag.
+        with transaction.atomic():
+            super().database_backwards(
+                app_label=app_label,
+                schema_editor=schema_editor,
+                from_state=from_state,
+                to_state=to_state,
+            )
+
+    def _validate(self) -> None:
+        if not isinstance(self.constraint, models.UniqueConstraint):
+            raise ValueError(
+                "SaferAddUniqueConstraint only supports the UniqueConstraint class"
+            )
+
+        if len(self._index_name) > MAX_RELATION_NAME_LEN:
+            raise ValueError(
+                f"Index name {self._index_name} is longer than "
+                f"{MAX_RELATION_NAME_LEN} characters limit."
+            )
```

Comment:
As noted above, we've computed `self._index_name` from the supplied constraint name; furthermore, the user (developer) has no insight into how we've done that.

This error message tells the user that the index name we've constructed is too long — but the real issue is that the supplied constraint name is longer than 59 characters, which is the arbitrary limit we've imposed by how we've calculated the index name.

I'm not sure the best way to fix this. Could we use the constraint name as the index name directly? Would that work? And would there be an issue? (Constraints and indexes live in separate namespaces in Postgres, so I think it should work.)

Reply:
Fixup: b3c627119240240411b7382bcf65c5c8a8896cdc

I had an issue when I tried reusing the index name. Not sure why I would go through such trouble otherwise. My shell output history says:

> psycopg.errors.DuplicateTable: relation "unique_int_field" already exists

I think it's just one of those cases of "tried this twice then got busy with something else before checking again".
Same name works:

```sql
-- test index name can be reused. Works on all version of Postgres supported.

CREATE TABLE example_table (
    id SERIAL PRIMARY KEY,
    int_field INT
);

CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS
unique_int_field ON example_table (int_field);

ALTER TABLE example_table
ADD CONSTRAINT unique_int_field UNIQUE USING INDEX unique_int_field;
```

Reply:
With the fixup, `MAX_RELATION_NAME_LEN` is now no longer used and can be removed.